### PR TITLE
Handle imports with same relative path

### PR DESCRIPTION
**Pull request by CraigSiemens**
- #1261

## Changes
- Identify merged includes using the full path to the file rather than just `filePath` which can be relative to the `basePath` and not unique.
- Added comments to the path properties to make it clearer what they contain.
- Other minor cleanup (renaming a parameter and making things private).

### CODE REVIEW

**Comment by CraigSiemens** [comment]
I wouldn't mind some feedback on whether tests should be added to this (I assume they should) and what they would look like. I think it'd either need a new fixture or an existing on to be updated but I don't know the preferred path.

**Comment by yonaskolb** [comment]
Thanks @CraigSiemens, some great changes. Does https://github.com/yonaskolb/XcodeGen/pull/1275 also solve your issue? If so perhaps we can merge that first, and you can add your documentation on top.

In terms of a test case, adding it to `Tests/Fixtures/paths_test.yml` is probably a good place

**Comment by CraigSiemens** [comment]
I tried it out and https://github.com/yonaskolb/XcodeGen/pull/1275 does fix the issue too.

I'm working on rebasing my branch onto master. I did find some improvements for the `cachedSpecFiles` that I'll add into this branch.

File: Sources/ProjectSpec/Project.swift [hunk 1]
```diff
@@ -155,14 +155,7 @@ extension Project: Equatable {
...
     public init(path: Path) throws {
-        var cachedSpecFiles: [Path: SpecFile] = [:]
-        let spec = try SpecFile(filePath: path, basePath: path.parent(), cachedSpecFiles: &cachedSpecFiles)
-        try self.init(spec: spec)
-    }
-
-    public init(path: Path, basePath: Path) throws {
-        var cachedSpecFiles: [Path: SpecFile] = [:]
-        let spec = try SpecFile(filePath: path, basePath: basePath, cachedSpecFiles: &cachedSpecFiles)
+        let spec = try SpecFile(path: path)
```
Comment (CraigSiemens) [inline Sources/ProjectSpec/Project.swift:158]: This is revering a change from https://github.com/yonaskolb/XcodeGen/pull/1275. Internally the init creates the `cachedSpecFiles` so callers don't have to make it.

File: Sources/ProjectSpec/SpecFile.swift [hunk 2]
```diff
@@ -63,26 +79,28 @@ public struct SpecFile {
...
+
         let includes = Include.parse(json: jsonDictionary["include"])
         let subSpecs: [SpecFile] = try includes
             .filter(\.enable)
             .map { include in
-                if let specFile = cachedSpecFiles[filePath] {
-                    return specFile
-                } else {
-                    return try SpecFile(include: include, basePath: basePath, relativePath: relativePath, cachedSpecFiles: &cachedSpecFiles, variables: variables)
-                }
```
Comment (CraigSiemens) [inline Sources/ProjectSpec/SpecFile.swift:81]: `filePath`, which is for the current spec, is not the correct key to use when looking for the `specFile` of an `include` that has it's own path.

**Comment by CraigSiemens** [comment]
@yonaskolb I believe I'm done rebasing/fixing so it's free for a re-review when you have a moment.

I've added in tests to `paths_test.yml` and verified they they failed before any of these changes were added to mater.

I also made a couple minor changes to the usage of `cachedSpecFiles`, I left a reason for some of them above.

**Comment by yonaskolb** [comment]
Thanks @CraigSiemens, I can hopefully get to this soon.

@ma-oli would you mind casting an eye over this too, as you were recently in this area.

**Comment by ma-oli** [comment]
I’m off right now until Monday but I’m happy to have a look when I’m back.

**Comment by ma-oli** [comment]
Hey; sorry about the delay. I'm looking at it now and I'm seeing some issues. I'll keep you posted.

**Comment by yonaskolb** [comment]
This issue has also come in https://github.com/yonaskolb/XcodeGen/issues/1299

File: Sources/ProjectSpec/SpecFile.swift [hunk 3]
```diff
@@ -48,8 +54,11 @@ public struct SpecFile {
         }
     }
 
-    public init(path: Path, cachedSpecFiles: inout [Path: SpecFile], variables: [String: String] = [:]) throws {
-        try self.init(filePath: path, basePath: path.parent(), cachedSpecFiles: &cachedSpecFiles, variables: variables)
+    public init(path: Path, basePath: Path? = nil, cachedSpecFiles: inout [Path: SpecFile], variables: [String: String] = [:]) throws {
+        let basePath = basePath ?? path.parent()
+        let filePath = try path.relativePath(from: basePath)
```
Comment (ma-oli) [inline Sources/ProjectSpec/SpecFile.swift:64]: Using a relative path ( "-r ." for instance) as a project root results in this line throwing an unmatchedAbsolutePath error. We might want to use basePath?.absolute(), either here or in SpecLoader.

**Review by ma-oli — CHANGES_REQUESTED** [review]
I think it's good but this breaks using a relative commit as a project root. Shouldn't be a big deal to fix I think.

File: Sources/ProjectSpec/SpecFile.swift (hunk 3 above)
Comment (CraigSiemens) [reply Sources/ProjectSpec/SpecFile.swift:64]: That's interesting, I'll try and take a look at that.

**Comment by CraigSiemens** [comment]
@ma-oli @yonaskolb I've added the fix for handling relative project paths.

**Review by ma-oli — APPROVED** [review]

**Comment by CraigSiemens** [comment]
@yonaskolb Any chance this could get merged since it's got an approval?

**Comment by yonaskolb** [comment]
> @yonaskolb Any chance this could get merged since it's got an approval?

Yep, could you please add a changelog entry. Also does this fix https://github.com/yonaskolb/XcodeGen/issues/1299?

**Comment by CraigSiemens** [comment]
Yeah, it should fix that issue. I've merged in master and updated the changelog.

**Review by yonaskolb — APPROVED** [review]
Fantastic, thanks for all your work on this @CraigSiemens 🙏

**Pull request merged**